x-pack/plugins/apm/public/components/shared/Links/APMLink.tsx

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { EuiLink, EuiLinkAnchorProps } from '@elastic/eui';
+import React from 'react';
+import url from 'url';
+import { useLocation } from '../../../hooks/useLocation';
+import { APMQueryParams, getSearchWithCurrentTimeRange } from './url_helpers';
+
+interface Props extends EuiLinkAnchorProps {
+  path?: string;
+  query?: APMQueryParams;
+}
+
+const APMLink: React.FC<Props> = ({ path, query, ...rest }) => {
+  const { search } = useLocation();
+  const nextSearch = getSearchWithCurrentTimeRange(search, query);
```

Comment:
I think `getSearchWithCurrentTimeRange` can be moved to APMLink - these are specific to APM, right?

Reply:
Same with `PERSISTENT_APM_PARAMS`, `PERSISTENT_APM_PARAMS` and `APMQueryParamsRaw`. I don't think these have a place in the generic link helpers. They are all very specific to APM links.

Reply:
Yes good point!

Reply:
I agree that `getSearchWithCurrentTimeRange` should move, but I don't think the `APMQueryParams` interfaces should move. Any url may need access to the current APM query params in order to construct a link to somewhere else, so those should stay somewhere central.

I'm on the fence about `PERSISTENT_APM_PARAMS` but I think those are pretty specific to APM so I'll move them, but then the `UpdateBreadcrumbs` component will just need to import them from `APMLink` ... which is probably ok.

Reply:
Instead of `getSearchWithCurrentTimeRange` we can simply read the time picker values from `urlParams`:

```js
  const { rangeFrom, rangeTo, refreshPaused, refreshInterval } = useUrlParams();
  const search = fromQuery({
    ...query,
    rangeFrom,
    rangeTo,
    refreshPaused,
    refreshInterval
  });
```

Again, I think this improves readability and avoids the "black box" problem that arises with some of our helpers.

Reply:
I've also seen this in a few places:

```js
const currentQuery = toQuery(currentSearch);
const nextQuery = {
  ...TIMEPICKER_DEFAULTS,
  ...pick(currentQuery, PERSISTENT_APM_PARAMS),
  ...query
};
```


I think we can replace that with:
```js
const persistentParams = getPersistentParams(urlParams)
```

where `getPersistentParams` is:

```ts
function getPersistentParams(urlParams: IUrlParams) {
  const { kuery, rangeFrom, rangeTo, refreshInterval, refreshPaused } = urlParams;
  return { kuery, rangeFrom, rangeTo, refreshInterval, refreshPaused };
}
```